File: game.jl
module Minesweeper

using Random: shuffle
using PaddedViews: PaddedView

export make_board, request, Index

Minefield = Array{Bool}
Index = CartesianIndex{2}

# Make a board with random mine positions
function make_board(h::Int,w::Int,n=0)::Minefield
    @assert n <= h * w

    b = zeros(Bool,(h,w))
    b[1:n] .= 1.0
    shuffle(b)
end

function neighbors(board::Minefield, ix::Index)::Int
    filt = ones(3,3)
    filt[2,2] = 0
    one = Index(1,1)

    pad = PaddedView( 0, board, (0:size(board,1)+1, 0:size(board,2)+1) )
    neighborhood = pad[ ix-one : ix+one ]

    sum(neighborhood .* filt)
end

function request(board::Minefield, ix::Tuple)::Int
    ix = Index(ix)
    if board[ix]
        # Game over
        return -1
    else
        return neighbors(board, ix)
    end
end

"""
Submit a grid of flags to check if the player has identified all the mines
correctly. Returns 1 if correct, and -1 if any mine is misidentified. In either
case, the game should end--with a win if correct and a failure if incorrect, so
that clients cannot repeatedly guess different configurations.
"""
function submit(board::Minefield, flags::Array{Bool})
    if all(board .== flags)
        return 1
    else
        # Game over
        return -1
    end
end

File: py/game.py
import numpy as np

class Minesweeper(object):

    def __init__(self, h, w, n):
        self.H = h
        self.W = w
        self.N = n

        self.board = self.make_board(h, w, n)
        self.game_over = False

    def make_board(self, h, w, n):
        vals = np.zeros(h * w, dtype=int)
        vals[:n] = 1
        # This runs in-place
        np.random.shuffle(vals)
        return vals.reshape((h,w))

    def get_neighbors(self, x, y):
        filt = np.ones((3,3))
        filt[1,1] = 0
        pad = np.pad(self.board,1)
        xypd = (x+1,y+1)
        tlpd = (x,y)
        brpd = (x+2,y+2)
        neighbors = (pad[x:x+3, y:y+3] * filt).sum() 

        return neighbors

    def request(self, r, c):
        if self.board[(r,c)] == 1:
            self.game_over = True
            print("Game over")
            return -1

        return self.get_neighbors(r, c)
    
    def submit(self, mask):
        if (mask == self.board).all():
            game_over = True
            print("You win")
            return 1
        else:
            game_over = True
            print("Loser")
            return -1
